### ISSUE TITLE
ArrayIsStoredDirectlyRule doesn't check for java primitives

### ISSUE THREAD

**Issue by frankholler** [pmd/pmd#5908]
**Affects PMD Version:** 
Any

**Rule:**

<https://pmd.github.io/pmd/pmd_rules_java_bestpractices.html#arrayisstoreddirectly>

**Description:**

private final byte[] data;
public demoDto(byte[] data) {
        this.data = data;
    }

**Expected outcome:**

PMD reports a violation at line, but that's wrong. That's a false positive.

This rule is correct for any Array containing regular java objects, but primitive types "int", "byte", "char", "boolean" ... may not be manipulated in any way as they ara always copied by value.

Please check java type of array that is an ancestor of "Object".

**Comment by oowekyala** [comment]
I don't think that's right... If you write
```java
class Foo { 
  byte[] data;  
  Foo(byte[] data) { this.data = data; }
}
int[] a = {1, 2, 3};
var foo = new Foo(a);
```
then `foo.data == a`, ie, they point to the same array instance. Then `a[0] = 6` will also change `foo.data[0]`, which the rule assumes is unexpected. This is what the rule is meant to flag, and that holds whatever the type of the elements of the array. What you describe would mean the rule is enforcing a deep copy of the array, which it cannot do.

**Comment by frankholler** [comment]
public class ByteArrayCopyTest {

    public class DemoDto {
        final byte[] data;
        public DemoDto(final byte[] data) {
            this.data = data;
        }

        public byte[] getData() {
            return data;
        }
    }
    @Test
    public void testManipulation() {
        byte[] in = new byte[] {
                (byte) 0x00, (byte) 0x01, (byte) 0x02, (byte) 0x03, (byte) 0x04, (byte) 0x05, (byte) 0x06, (byte) 0x07, (byte) 0x08,
                (byte) 0x09, (byte) 0x0a, (byte) 0x0b, (byte) 0x0c, (byte) 0x0d, (byte) 0x0e, (byte) 0x0f };

        DemoDto demoDto = new DemoDto(in);
        byte[] test = demoDto.getData();
        test[0] = (byte) 0x09;
        Assert.assertEquals((byte) 0x00, in[0]);
    }
}

my fault, your right, i only thought about the length itself not the replacement of the value.
